pallets/ddc-validator/src/lib.rs

```diff
@@ -0,0 +1,881 @@
+					(cdn_node, validation_decision.totals.sent)
+				})
+				.collect();
+
+			// Store CDN node reward points on-chain
+			let signer: Signer<T, T::AuthorityId> = Signer::<_, _>::any_account();
+			if !signer.can_sign() {
+				log::warn!("No local accounts available to set era reward points. Consider adding one via `author_insertKey` RPC.");
+				return
+			}
+			// ToDo: replace local call by a call from `ddc-staking` pallet
+			let _tx_res = signer.send_signed_transaction(|_account| Call::set_era_reward_points {
+				era: current_era - 1,
+				stakers_points: cdn_nodes_reward_points.clone(),
+			});
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		<T as frame_system::Config>::AccountId: AsRef<[u8]> + UncheckedFrom<T::Hash>,
+		<BalanceOf<T> as HasCompact>::Type: Clone + Eq + PartialEq + Debug + TypeInfo + Encode,
+	{
+		/// Run a process at the same time on all the validators.
+		#[pallet::weight(10_000)]
+		pub fn send_signal(origin: OriginFor<T>) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			Signal::<T>::set(Some(true));
+
+			Ok(())
+		}
+
+		/// Set validation decision for a given CDN node in an era.
+		#[pallet::weight(10_000)]
+		pub fn set_validation_decision(
+			origin: OriginFor<T>,
+			era: EraIndex,
+			cdn_node: T::AccountId,
+			validation_decision: ValidationDecision,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			// ToDo: check if origin is a validator.
+			// ToDo: check if the era is current - 1.
+			// ToDo: check if the validation decision is not set yet.
+			// ToDo: check cdn_node is known to ddc-staking.
+
+			ValidationDecisions::<T>::insert(era, cdn_node, validation_decision);
+
+			// ToDo: emit event.
+
+			Ok(())
+		}
+
+		/// Set reward points for CDN participants at the given era.
+		///
+		/// ToDo: remove it when the off-chain worker will be able to set reward points using the
+		/// same call defined in `pallet-ddc-staking`.
+		///
+		/// `stakers_points` is a vector of (stash account ID, reward points) pairs. The rewards
+		/// distribution will be based on total reward points, with each CDN participant receiving a
+		/// proportionate reward based on their individual reward points.
+		///
+		/// See also  [`pallet_ddc_staking::ErasEdgesRewardPoints`].
+		#[pallet::weight(100_000)]
+		pub fn set_era_reward_points(
+			origin: OriginFor<T>,
+			era: EraIndex,
+			stakers_points: Vec<(T::AccountId, u64)>,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			<ddc_staking::pallet::ErasEdgesRewardPoints<T>>::mutate(era, |era_rewards| {
+				for (staker, points) in stakers_points.into_iter() {
+					*era_rewards.individual.entry(staker).or_default() += points;
+					era_rewards.total += points;
+				}
+			});
+
+			Ok(())
+		}
+
+		#[pallet::weight(100_000)]
+		pub fn charge_payments_cdn(
+			origin: OriginFor<T>,
+			paying_accounts: Vec<BucketsDetails<ddc_accounts::BalanceOf<T>>>,
+		) -> DispatchResult {
+			let controller = ensure_signed(origin)?;
```

Comment:
What is the meaning of `controller` in this context?

Reply:
this is actually a registered OCW key by validator

Reply:
With the presence of tight coupling with staking-related stuff there this name is confusing.